visitors/decorators.py

```
@@ -103,6 +103,9 @@ def inner(*args: Any, **kwargs: Any) -> HttpResponse:
             response = view_func(*args, **kwargs)
             if log_visit:
                 VisitorLog.objects.create_log(request, response.status_code)
+            Visitor.objects.increment_uses(request)
+            if not request.visitor.has_uses_remaining:
+                raise PermissionDenied(_("Visitor access denied (usage exhausted)."))
```

I was going to leave this out and just let the middleware take care of validating, but I felt a bit more secure having this check here and a test with it

---

I have a question about this - given that the decorator is not guarded by `transaction.atomic`, if this exception is raised, we will also have a.) incremented the uses, and b.) logged the visit. I'm not sure whether this is the correct behaviour or not. On the plus side, we're logging all of the "bad" uses, which may be useful for debugging; on the minus side, it's logging all of the "bad" uses, which is just using up space needlessly - may be better to just fail hard (which is our usual philosophy). 🤔

---

I'd thought that incrementing the uses would be ok since the validation would then start failing in the middleware and that would stop the incrementing. I think you're right about the logging though maybe it should go after the error.

---

something like this perhaps ...

```python
with transaction.atomic():
    Visitor.objects.increment_uses(request)
    if not request.visitor.has_uses_remaining:
        raise PermissionDenied(_("Visitor access denied (usage exhausted)."))
    if log_visit:
        VisitorLog.objects.create_log(request, response.status_code)
```

---

I think the simplest solution would be to have the call to `validate` here in the decorator, not the middleware, and then let the middleware handle any errors raised. In this current implementation we are splitting responsibilities across middleware / decorator - both perform validation - some different, some the same (duplicated). It would be cleaner if middleware was responsible for fetching, and the decorator for validating.

---

possibly it's work for another PR to refactor the code and update the tests since the Middleware (`VisitorRequestMiddleware`) already contained the validation. What I've done in the past is setup a chore that gets merged into develop first and then into the feature branch and then the feature branch is updated accordingly. It keeps the PRs small and focussed on one task.